org.eclipse.jdt.core/model/org/eclipse/jdt/internal/core/JavaModelManager.java

```diff
@@ -1357,34 +1357,31 @@ public static class PerProjectInfo {
 
 		static final IJavaModelStatus NEED_RESOLUTION = new JavaModelStatus();
 
-		public IProject project;
-		public Object savedState;
-		public boolean triedRead;
-		public IClasspathEntry[] rawClasspath;
-		public IClasspathEntry[] referencedEntries;
-		public IJavaModelStatus rawClasspathStatus;
-		public int rawTimeStamp = 0;
-		public boolean writtingRawClasspath = false;
-		public IClasspathEntry[] resolvedClasspath;
-		public IJavaModelStatus unresolvedEntryStatus;
-		public Map<IPath, IClasspathEntry> rootPathToRawEntries; // reverse map from a package fragment root's path to the raw entry
-		public Map<IPath, IClasspathEntry> rootPathToResolvedEntries; // map from a package fragment root's path to the resolved entry
-		public IPath outputLocation;
-		public Map<IPath, ObjectVector> jrtRoots; // A map between a JRT file system (as a string) and the package fragment roots found in it.
-
-		public IEclipsePreferences preferences;
-		public Hashtable<String, String> options;
+		public final IProject project;
+		public volatile Object savedState;
+		public volatile boolean triedRead;
+		public volatile IClasspathEntry[] rawClasspath;
+		public volatile IClasspathEntry[] referencedEntries;
+		public volatile IJavaModelStatus rawClasspathStatus;
+		public volatile int rawTimeStamp;
+		public volatile boolean writtingRawClasspath;
+		public volatile IClasspathEntry[] resolvedClasspath;
+		public volatile IJavaModelStatus unresolvedEntryStatus;
+		public volatile Map<IPath, IClasspathEntry> rootPathToRawEntries; // reverse map from a package fragment root's path to the raw entry
+		private Map<IPath, IClasspathEntry> rootPathToResolvedEntries; // map from a package fragment root's path to the resolved entry
```

Comment:
This change is problematic for XBase.

Reply:
I've answered on https://github.com/eclipse/Xpect/issues/353. 

I believe there is no need for Xtext to use **this** part of internal JDT code, there is another part of internal JDT code that should be better used.

Reply:
@iloveeclipse and what would the proper api to be used be.

Reply:
`PackageFragmentRoot.getResolvedClasspathEntry()`.